src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs

```diff
@@ -120,171 +110,176 @@ public void Dispose()
 
         private partial void DisposeWorker();
 
-        internal bool IsAnalyzerDependencyPath(string fullPath)
+        public void AddDependencyLocation(string originalPath)
         {
             CheckIfDisposed();
 
-            lock (_guard)
+            CompilerPathUtilities.RequireAbsolutePath(originalPath, nameof(originalPath));
+            var simpleName = PathUtilities.GetFileName(originalPath, includeExtension: false);
+            string realPath = originalPath;
+            IAnalyzerPathResolver? resolver = null;
+            foreach (var current in AnalyzerPathResolvers)
             {
-                return _analyzerAssemblyInfoMap.ContainsKey(fullPath);
+                if (current.IsAnalyzerPathHandled(originalPath))
+                {
+                    resolver = current;
+                    realPath = resolver.GetRealAnalyzerPath(originalPath);
+                    break;
+                }
             }
-        }
-
-        public void AddDependencyLocation(string fullPath)
-        {
-            CheckIfDisposed();
 
-            CompilerPathUtilities.RequireAbsolutePath(fullPath, nameof(fullPath));
-            string simpleName = PathUtilities.GetFileName(fullPath, includeExtension: false);
+            var assemblyName = readAssemblyName(realPath);
 
             lock (_guard)
             {
-                if (!_knownAssemblyPathsBySimpleName.TryGetValue(simpleName, out var paths))
+                if (_originalPathInfoMap.TryAdd(originalPath, (resolver, realPath, assemblyName)))
+                {
+                    // In the case multiple original paths map to the same real path then the first on
+                    // wins.
+                    //
+                    // An example reason to map multiple original paths to the same real path would be to
+                    // unify references.
+                    _ = _realToOriginalPathMap.TryAdd(realPath, originalPath);
+
+                    if (!_assemblySimpleNameToOriginalPathListMap.TryGetValue(simpleName, out var set))
+                    {
+                        set = new();
+                        _assemblySimpleNameToOriginalPathListMap[simpleName] = set;
```

Comment:
fyi, we have a MultiDictionary which already has hte semantics of mapping a key to a set of values.